### PR TITLE
Optional command line flags (--exportInvalidMigrations & --customRecipe)

### PR DESCRIPTION
**Pull request by TomiwaOke**
- Allow for users to specify custom recipes to use for migration
- Add lenient export flag which will include invalid migrations in newly created text files after --file type migrations
- Show users which recipes were used for migrations after migrations have been performed

### CODE REVIEW

File: migrator/src/main/java/com/google/phonenumbers/migrator/CommandLineMain.java [hunk 1]
```diff
@@ -56,6 +59,24 @@
...
   String countryCode;
 
+  @ArgGroup()
+  OptionalParameterType optionalParameter;
+
+  static class OptionalParameterType {
+    @Option(names = {"-l", "--lenientExport"},
+        description = "boolean flag specifying that text files created after the migration process"
+            + " for standard recipe --file migrations should contain the migrated version of a given"
+            + " phone number when a migration has taken place on it, regardless of whether the"
```
Comment (brittanyroesch) [inline migrator/src/main/java/com/google/phonenumbers/migrator/CommandLineMain.java]: nit: maybe remove this part 'when a migration has taken place on it' to shorten it

File: migrator/src/main/java/com/google/phonenumbers/migrator/CommandLineMain.java [hunk 2]
```diff
@@ -56,6 +59,24 @@
...
+            + " migration resulted in an invalid phone number. By default, a strict approach is used"
+            + " and when a migration is seen as invalid, the original phone number is written to file.")
```
Comment (brittanyroesch) [inline migrator/src/main/java/com/google/phonenumbers/migrator/CommandLineMain.java]: Maybe add that any invalid numbers are printed out afterwards so the user is aware which numbers were invalid

File: migrator/src/main/java/com/google/phonenumbers/migrator/MigrationFactory.java [hunk 3]
```diff
@@ -56,21 +56,23 @@ public static MigrationJob createMigration(String number, String country) throws
         .orElseThrow(() -> new RuntimeException(
             "Country code " + countryCode+ " not supported in metadata"));
 
-    return new MigrationJob(numberRanges, countryCode, recipes, ranges);
+    return new MigrationJob(numberRanges, countryCode, recipes, ranges, false);
```
Comment (brittanyroesch) [inline migrator/src/main/java/com/google/phonenumbers/migrator/MigrationFactory.java]: When having a parameter that isn't clear, such as false, it is always good to add an inline comment, ie:
return new MigrationJob(numberRanges, countryCode, recipes, ranges, /* parameterName= */ false);

File: migrator/src/main/java/com/google/phonenumbers/migrator/MigrationJob.java [hunk 4]
```diff
@@ -44,15 +46,18 @@
   private final CsvTable<RangeKey> recipesTable;
   private final ImmutableList<MigrationEntry> migrationEntries;
   private final DigitSequence countryCode;
+  private final boolean lenientExport;
```
Comment (brittanyroesch) [inline migrator/src/main/java/com/google/phonenumbers/migrator/MigrationJob.java]: Could you add a comment above this as it isn't obivous when looking at this file what it is.

File: migrator/src/test/java/com/google/phonenumbers/migrator/CommandLineMainTest.java [hunk 5]
```diff
@@ -73,4 +73,17 @@ public void createFromPath_expectSufficientArguments() {
     assertThat(p.numberInput.file).matches(TEST_FILE_INPUT);
     assertThat(p.numberInput.number).isNull();
   }
+
+  @Test
+  public void createMigrationJob_lenientFlagAndCustomRecipe_expectException() {
```
Comment (brittanyroesch) [inline migrator/src/test/java/com/google/phonenumbers/migrator/CommandLineMainTest.java]: Could you make this clear in the arg docs that these specific args cannot be used together

File: migrator/src/main/java/com/google/phonenumbers/migrator/CommandLineMain.java [hunk 6]
```diff
@@ -56,6 +59,26 @@
...
+  @ArgGroup()
+  OptionalParameterType optionalParameter;
+
+  static class OptionalParameterType {
+    @Option(names = {"-l", "--lenientExport"},
+        description = "boolean flag specifying that text files created after the migration process"
+            + " for standard recipe --file migrations should contain the migrated version of a given"
+            + " phone number, regardless of whether the migration resulted in an invalid phone number."
+            + " By default, a strict approach is used and when a migration is seen as invalid, the"
+            + " original phone number is written to file. Invalid numbers will be printed at the"
```
Comment (brittanyroesch) [inline migrator/src/main/java/com/google/phonenumbers/migrator/CommandLineMain.java:71]: I think tweaking this just a bit to be more clear, that all migrated numbers will be in their original order while migrated invalid numbers will be added to to bottom of the text file.

I also think we should double check with David on this, maybe you can ask him the next time you talk to him if this seems reasonable depending on the different use cases.

File: migrator/src/main/java/com/google/phonenumbers/migrator/MigrationFactory.java [hunk 7]
```diff
@@ -56,21 +56,24 @@ public static MigrationJob createMigration(String number, String country) throws
         .orElseThrow(() -> new RuntimeException(
             "Country code " + countryCode+ " not supported in metadata"));
 
-    return new MigrationJob(numberRanges, countryCode, recipes, ranges);
+    return new MigrationJob(numberRanges, countryCode, recipes, ranges, /*lenientExport=*/false);
```
Comment (brittanyroesch) [inline migrator/src/main/java/com/google/phonenumbers/migrator/MigrationFactory.java]: You can add a space between the first star, the equals and the last slash, ie. 

/* lenientExport= */ false

**Review by brittanyroesch — APPROVED** [review]
LGTM, will wait for Michael to take a look.

File: migrator/src/main/java/com/google/phonenumbers/migrator/CommandLineMain.java [hunk 8]
```diff
@@ -56,6 +59,26 @@
       description = "The BCP-47 country code the given phone number(s) belong to (e.g. 44)")
   String countryCode;
 
+  @ArgGroup()
+  OptionalParameterType optionalParameter;
+
+  static class OptionalParameterType {
+    @Option(names = {"-l", "--lenientExport"},
```
Comment (gonecoding) [inline migrator/src/main/java/com/google/phonenumbers/migrator/CommandLineMain.java]: How about being more explicit with the parameter name so it's a bit more self-explanatory, e. g. `--allowInvalidPhoneNumbers`?

File: migrator/src/main/java/com/google/phonenumbers/migrator/MigrationFactory.java (hunk 7 above)
Comment (gonecoding) [reply migrator/src/main/java/com/google/phonenumbers/migrator/MigrationFactory.java]: Here and below

File: migrator/src/main/java/com/google/phonenumbers/migrator/MigrationJob.java (hunk 4 above)
Comment (gonecoding) [reply migrator/src/main/java/com/google/phonenumbers/migrator/MigrationJob.java]: See my above comment about renaming this parameter to be more self-explanatory

**Review by gonecoding — APPROVED** [review]

**Pull request merged**